File: src/rust/tracer/ftrace.rs
use object::{Object, ObjectSymbol, SymbolKind};
use std::{collections::HashMap, fmt};

use crate::tracer::ringbuf::RingBuf;

const FTRACE_CAPACITY: usize = 32;

pub struct FuncTracer {
    pub ring_buf: RingBuf<FTraceEntry>,
    pub symtab: HashMap<u64, String>,
    depth: u32,
}

impl FuncTracer {
    pub fn new(elf_path: &std::path::Path) -> Self {
        let data = std::fs::read(elf_path).expect("failed to read ELF");
        let obj = object::File::parse(&*data).expect("failed to parse ELF");
        let mut symtab = HashMap::new();
        for sym in obj.symbols() {
            if sym.kind() == SymbolKind::Text && sym.size() > 0 {
                if let Ok(name) = sym.name() {
                    symtab.insert(sym.address(), name.to_string());
                }
            }
        }
        Self {
            ring_buf: RingBuf::new(FTRACE_CAPACITY),
            symtab,
            depth: 0,
        }
    }

    pub fn push_call(&mut self, pc: u64, dnpc: u64, disasm: &str) {
        let func_name = self
            .symtab
            .get(&dnpc)
            .cloned()
            .unwrap_or_else(|| format!("{dnpc:#018x}"));
        let entry = FTraceEntry::new(pc, dnpc, self.depth, FuncType::Call(func_name), disasm);
        self.ring_buf.push(entry);
        self.depth += 1;
    }

    pub fn push_ret(&mut self, pc: u64, dnpc: u64, disasm: &str) {
        self.depth = self.depth.saturating_sub(1);
        let entry = FTraceEntry::new(pc, dnpc, self.depth, FuncType::Ret, disasm);
        self.ring_buf.push(entry);
    }
}

pub enum FuncType {
    Call(String),
    Ret,
}

pub struct FTraceEntry {
    pub pc: u64,
    pub dnpc: u64,
    pub depth: u32,
    pub func_type: FuncType,
    pub disasm: String,
}

impl FTraceEntry {
    pub fn new(pc: u64, dnpc: u64, depth: u32, func_type: FuncType, disasm: &str) -> Self {
        Self {
            pc,
            dnpc,
            depth,
            func_type,
            disasm: disasm.to_string(),
        }
    }
}

impl fmt::Display for FTraceEntry {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let indent = "  ".repeat(self.depth as usize);
        match &self.func_type {
            FuncType::Call(name) => {
                write!(
                    f,
                    "{:#018x}: {indent}call [{name}@{:#018x}] ({})",
                    self.pc, self.dnpc, self.disasm
                )
            }
            FuncType::Ret => {
                write!(f, "{:#018x}: {indent}ret ({})", self.pc, self.disasm)
            }
        }
    }
}

File: src/cpp/tracer/ftrace.cc
#include "tracer/ftrace.h"

#include <fstream>
#include <vector>

#include "absl/log/log.h"
#include "absl/strings/str_format.h"
#include "elfio/elfio.hpp"

namespace npc {

FuncTracer::FuncTracer(const std::filesystem::path& elf_path)
    : ring_buf(kFtraceCapacity) {
  ELFIO::elfio reader;
  if (!reader.load(elf_path.string())) {
    LOG(ERROR) << "Failed to load ELF: " << elf_path;
    return;
  }

  for (const auto& section : reader.sections) {
    if (section->get_type() != ELFIO::SHT_SYMTAB &&
        section->get_type() != ELFIO::SHT_DYNSYM) {
      continue;
    }
    ELFIO::symbol_section_accessor symbols(reader, section.get());
    for (ELFIO::Elf_Xword i = 0; i < symbols.get_symbols_num(); ++i) {
      std::string name;
      ELFIO::Elf64_Addr value = 0;
      ELFIO::Elf_Xword size = 0;
      unsigned char bind = 0, type = 0, other = 0;
      ELFIO::Elf_Half section_index = 0;
      symbols.get_symbol(i, name, value, size, bind, type, section_index,
                         other);
      if (type == ELFIO::STT_FUNC && size > 0 && !name.empty()) {
        symtab[value] = name;
      }
    }
  }
}

void FuncTracer::push_call(uint64_t pc, uint64_t dnpc,
                            const std::string& disasm) {
  auto it = symtab.find(dnpc);
  std::string func_name =
      (it != symtab.end()) ? it->second
                           : absl::StrFormat("0x%016x", dnpc);
  ring_buf.push(FTraceEntry(pc, dnpc, depth_, FuncType::kCall,
                             std::move(func_name), disasm));
  ++depth_;
}

void FuncTracer::push_ret(uint64_t pc, uint64_t dnpc,
                           const std::string& disasm) {
  if (depth_ > 0) --depth_;
  ring_buf.push(
      FTraceEntry(pc, dnpc, depth_, FuncType::kRet, "", disasm));
}

}  // namespace npc
